libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFSDKEncryptedURLCache.m

```diff
@@ -0,0 +1,121 @@
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "SFSDKEncryptedURLCache.h"
+#import "SFKeyStoreManager.h"
+#import "NSData+SFAdditions.h"
+
+static NSString * const kURLSchemePrefix = @"sfsdkURLCache://";
+static NSString * const kURLCacheEncryptionKeyLabel = @"com.salesforce.URLCache.encryptionKey";
+
+@interface SFSDKEncryptedURLCache()
+
+@property SFEncryptionKey *encryptionKey;
+
+@end
+
+@implementation SFSDKEncryptedURLCache
+
+- (instancetype)initWithMemoryCapacity:(NSUInteger)memoryCapacity
+                          diskCapacity:(NSUInteger)diskCapacity
+                              diskPath:(NSString *)path {
+    self = [super initWithMemoryCapacity:memoryCapacity diskCapacity:diskCapacity diskPath:path];
+    if (self) {
+        _encryptionKey = [[SFKeyStoreManager sharedInstance] retrieveKeyWithLabel:kURLCacheEncryptionKeyLabel autoCreate:YES];
+    }
+    return self;
+}
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        _encryptionKey = [[SFKeyStoreManager sharedInstance] retrieveKeyWithLabel:kURLCacheEncryptionKeyLabel autoCreate:YES];
+    }
+    return self;
+}
+
+- (nullable NSCachedURLResponse *)cachedResponseForRequest:(NSURLRequest *)request {
+    // For request.URL
+    NSURLRequest *requestWithSecureURL = [self requestWithSecureURLForRequest:request];
+    if (!requestWithSecureURL) {
+        [SFSDKCoreLogger d:[self class] format:@"RequestWithSecureURL is nil, unable to check cache for request"];
+        return nil;
+    }
+
+    // For response.data
+    NSCachedURLResponse *cachedResponse = [super cachedResponseForRequest:requestWithSecureURL];
+    if (cachedResponse) {
+        NSData *decryptedResponseData = [self.encryptionKey decryptData:cachedResponse.data];
+        if (decryptedResponseData) {
+            NSCachedURLResponse *decryptedURLResponse = [[NSCachedURLResponse alloc] initWithResponse:cachedResponse.response data:decryptedResponseData userInfo:cachedResponse.userInfo storagePolicy:cachedResponse.storagePolicy];
+            return decryptedURLResponse;
+        } else {
+             [SFSDKCoreLogger d:[self class] format:@"Unable to decrypt response"];
+        }
+    }
+    return nil;
+}
+
+- (void)storeCachedResponse:(NSCachedURLResponse *)cachedResponse
+                 forRequest:(NSURLRequest *)request {
+    // For request.URL
+    NSURLRequest *requestWithSecureURL = [self requestWithSecureURLForRequest:request];
+    if (!requestWithSecureURL) {
+        [SFSDKCoreLogger d:[self class] format:@"RequestWithSecureURL is nil"];
+        return;
+    }
+    
+    // For cachedResponse.data
+    NSData *encryptedResponseData = [self.encryptionKey encryptData:cachedResponse.data];
+    if (!encryptedResponseData) {
+        [SFSDKCoreLogger d:[self class] format:@"Unable to encrypt response"];
```

Comment:
Should log level be d(ebug) or e(rror)?

Reply:
👍I'll change to error. Also, what do you think about trying to remove the item from the cache if decrypting fails?

Reply:
You mean in `cachedResponseForRequest`? If the cache returns nil, wouldn't the request be sent to the server and the server's response be written in the cache (over the one we couldn't decrypt)?